kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -0,0 +1,915 @@
+		svc.Status.LoadBalancer.Ingress = svcIngress
+
+		_, err = c.clientSet.CoreV1().Services(svc.Namespace).UpdateStatus(context.Background(), svc, metav1.UpdateOptions{})
+		if err != nil {
+			log.WithError(err).Errorf("Failed to update service status %s/%s", svc.Namespace, svc.Name)
+			return err
+		}
+	}
+	return nil
+}
+
+// assignIP tries to assign IP address for Service.
+func (c *loadBalancerController) assignIP(svc *v1.Service) ([]string, error) {
+	if len(c.ipPools) == 0 {
+		return nil, nil
+	}
+
+	handle, err := createHandle(svc)
+	if err != nil {
+		return nil, err
+	}
+
+	loadBalancerIps, ipv4Pools, ipv6Pools, err := c.parseAnnotations(svc.Annotations)
+	if err != nil {
+		return nil, err
+	}
+
+	var assignedIPs []string
+
+	metadataAttrs := map[string]string{
+		ipam.AttributeService:   svc.Name,
+		ipam.AttributeNamespace: svc.Namespace,
+		ipam.AttributeType:      string(svc.Spec.Type),
+		ipam.AttributeTimestamp: time.Now().UTC().String(),
+	}
+
+	namespacedName := types.NamespacedName{
+		Name:      svc.Name,
+		Namespace: svc.Namespace,
+	}
+
+	if loadBalancerIps != nil {
+		// User requested specific IP, attempt to allocate
+		for _, addrs := range loadBalancerIps {
+			skipAddrs := false
+			for lbingress := range c.allocationTracker.ipsByService[namespacedName] {
+				// We must be trying to assign missing address due to an error,
+				// skip this assignment as it's already assigned and move onto the next one
+				if lbingress == addrs.String() {
+					skipAddrs = true
+				}
+			}
+
+			if skipAddrs {
+				continue
+			}
+
+			ipamArgs := ipam.AssignIPArgs{
+				IP:          addrs,
+				Hostname:    api.VirtualLoadBalancer,
+				HandleID:    &handle,
+				Attrs:       metadataAttrs,
+				IntendedUse: api.IPPoolAllowedUseLoadBalancer,
+			}
+
+			err = c.calicoClient.IPAM().AssignIP(context.Background(), ipamArgs)
+			if err != nil {
+				log.WithField("ip", addrs).WithError(err).Warn("failed to assign ip to node")
+				return nil, err
+			}
+			assignedIPs = append(assignedIPs, addrs.String())
+			c.allocationTracker.assignAddressToService(namespacedName, addrs.String())
+		}
+		return assignedIPs, err
+	}
+
+	// Build AssignArgs based on Service IP family attr
+	num4 := 0
+	num6 := 0
+
+	for _, ipFamily := range svc.Spec.IPFamilies {
+		if ipFamily == v1.IPv4Protocol {
+			num4++
+		}
+		if ipFamily == v1.IPv6Protocol {
+			num6++
+		}
+	}
+
+	// Check if IP from ipfamily is already assigned, skip it as we're trying to assign only the missing one.
+	// This can happen when error happened during the initial assignment, and now we're trying to assign ip again from the syncIPAM func
+	for ingress := range c.allocationTracker.ipsByService[namespacedName] {
+		if ip := cnet.ParseIP(ingress); ip != nil {
+			if ip.To4() != nil {
+				num4 = 0
+			}
+			if ip.To16() != nil {
+				num6 = 0
+			}
+		}
+	}
+
+	if num4 == 0 && num6 == 0 {
+		return nil, errors.New("No new IPs to assign, Service already has ipv4 and ipv6 address")
```

Comment:
I think this error is not worded correctly - this could happen if the Service is IPv4 only and already has an IPv4 address, for example.

```suggestion
		return nil, errors.New("No new IPs to assign, Service already has desired LB addresses")
```